test/plugins/csv_output/test_csv_output_plugin.py

```diff
@@ -0,0 +1,239 @@
+    arg_headers = [MagicMock(), MagicMock()]
+    csv_out = CSV_Output(arg_name, arg_headers)
+
+    example = {
+        'vehicle_rep':{
+            'plugin_1': ['1', '2', '3']
+        },
+        'learning_system':{
+            'plugin_2': ['a', 'b', 'c'],
+            'plugin_3': ['x', 'y', 'z']
+        },
+        'planning_system':{}
+    }
+    expected_buffer = ['1', '2', '3', 'a', 'b', 'c', 'x' ,'y', 'z']
+
+    csv_out.update(low_level_data= [], high_level_data = example)
+
+    assert csv_out.current_buffer == expected_buffer
+
+def test_render_reasoning_creates_expected_file_on_first_frame():
+    arg_name = MagicMock()
+    arg_headers = ['header1', 'header2']
+    csv_out = CSV_Output(arg_name, arg_headers)
+    csv_out.filename_preamble = 'test_'
+
+    csv_out.render_reasoning()
+
+    assert csv_out.file_name.startswith(csv_out.filename_preamble)
+    assert os.path.exists(csv_out.file_name)
+
+    # cleanup
+    if os.path.exists(csv_out.file_name):
+        os.remove(csv_out.file_name)
+
+def test_render_reasoning_does_not_create_new_file_when_not_first_frame():
+    arg_name = MagicMock()
+    arg_headers = ['header1', 'header2']
+    csv_out = CSV_Output(arg_name, arg_headers)
+    csv_out.filename_preamble = 'test_'
+
+    # test specific conditions
+    target_file_name = csv_out.filename_preamble + "render_reasoning_not_first_frame" + ".csv"
+    csv_out.file_name = target_file_name
+    csv_out.first_frame = False
+
+    # create file first so csv_out pluging has something to write to
+    with open(target_file_name, 'a') as file: 
+        delimiter = ','
+        file.write(delimiter.join(csv_out.headers) + '\n')
+
+    # act
+    csv_out.render_reasoning()    
+    
+    try:
+        # if the current file name attribute doesn't match the original target file name
+        # then a new file was created
+        assert(csv_out.file_name == target_file_name)
+
+    except:
+        os.remove(csv_out.file_name)
+        raise AssertionError("Unexpected new file created after first frame")
+    
+    finally:
+        os.remove(target_file_name)
+
+def test_render_reasoning_changes_file_when_lines_per_file_reached():
+    arg_name = MagicMock()
+    arg_headers = ['header1', 'header2']
+    csv_out = CSV_Output(arg_name, arg_headers)
+    csv_out.filename_preamble = 'test_'
+
+    # test specific conditions
+    old_file_name = csv_out.filename_preamble + "render_reasoning_end_of_file" + ".csv"
+    csv_out.file_name = old_file_name
+    csv_out.first_frame = False
+    csv_out.lines_per_file = 1
+    csv_out.lines_current = 0
+
+    # act
+    csv_out.render_reasoning()
+
+    # check if a new file was set, but NOT yet created
+    try:
+        assert csv_out.file_name != old_file_name
+        assert not os.path.exists(csv_out.file_name)
+    except:
+        raise AssertionError("file name was not updated or new file was created prematurely")
+    finally:
+        os.remove(old_file_name)
```

Comment:
Small nitpick: please end files with a newline. That way when someone cats the file in a terminal it looks nicer.